File: Week 0/RippleCarryAdder-4-bit/Code.jl
function AddBinaryIntegers(a,b,n,max)
    c = a+b
    carry = 0
    if c > max
        c = c - max - 1
        carry = 1
    end
    return c, carry
end

n = 8
max = 2^n - 1

DecimalNumbers = [i for i in 0:max]

open("tracefile.txt","w") do file
    for i in DecimalNumbers
        for j in DecimalNumbers
            I1 = string(i, base = 2, pad=n)             # Convert i to binary string
            I2 = string(j, base = 2, pad=n)             # Convert j to binary string
            Output, OutputCarry = AddBinaryIntegers(i,j,n,max) # Function returns output and carry in Decimal
            O, C = string(Output, base = 2, pad=n), string(OutputCarry, base = 2, pad=1) # Calculating output and carry in Binary
            M = "11111"                                 # Mask bit
            write(file,"$I1$I2 $C$O $M\n")              # Writing to file
        end
    end
end

File: Week 1/PrimeCheckerAdder/Code.py
import numpy as np
import sympy as sp

n = 4
Max = 2**n

def generateBinaryStrings(n, Max):
    return [bin(i)[2:].zfill(n) for i in range(Max)]

BinaryStrings = generateBinaryStrings(n, Max)

def addBinaryIntegers(a,b,n, Max):
    c = a + b
    carry = 0
    if c > Max - 1:
        c = c - Max
        carry = 1
    return c, carry

def subBinaryIntegers(a,b,n, Max):
    b = 2**n - b
    c = a + b
    carry = 0
    if c > Max - 1:
        c = c - Max
        carry = 1
    carry = 1 - carry
    return c, carry

subBinaryIntegers(2,14,n,Max)

def PrimeCheck(string):
    if (sp.isprime(int(string,2))):
        return 1
    else:
        return 0

numberOfOutputBits = 7

with open("Tracefile.txt", "w") as file: 
    for i in range(Max):
        for j in range(Max):
            I1 = BinaryStrings[i]
            I2 = BinaryStrings[j]
            opa = PrimeCheck(I1)
            opb = PrimeCheck(I2)
            if opa == 1 and opb == 1:
                output, carry = subBinaryIntegers(i,j,n, Max)
            elif opa == 1 or opb == 1:
                output, carry = addBinaryIntegers(i,j,n, Max)
            else:
                output, carry = 0, 0
            mask = '1'*numberOfOutputBits
            o = bin(output)[2:].zfill(n)
            str = "{}{} {}{}{}{} {}\n".format(I1, I2, opa, opb, carry, o, mask)
            file.write(str)